### PR TITLE
Incorporate the Idemopotency support into the Kinesis Input Operator.

### PR DESCRIPTION
**Pull request by chaithu14**

### CODE REVIEW

**Comment by tweise** [comment]
Why are we storing the messages in recovery state when they are available from Kinesis. We should just keep track of the offsets.

**Comment by chaithu14** [comment]
I think, it would be redundant process, If we store the offsets in recovery state. Because the replay method, has to read those records from kinesis  and convert those records to tuples again.

**Comment by tweise** [comment]
Currently everything is spooled to HDFS for the recovery case. Instead we want to optimize regular processing, not the exception case.

File: contrib/src/main/java/com/datatorrent/contrib/kinesis/AbstractKinesisOutputOperator.java [hunk 1]
```diff
@@ -54,10 +54,10 @@
   @NotNull
   private String secretKey;
   protected static transient AmazonKinesisClient client = null;
+  private transient KinesisJsonEncoder encoder;
```
Comment (tweise) [inline contrib/src/main/java/com/datatorrent/contrib/kinesis/AbstractKinesisOutputOperator.java]: Encoder needs to be abstract. Cannot assume JSON, but it is a good default.

**Comment by chaithu14** [comment]
We will fetch the data from KinesisInputOperator as follows:
    * Records would be read from multiple shards.
    * There is one thread created to service each shard.
    * This thread continuously keeps fetching records from the shard it is servicing.
    * Fetched records are stored into the buffer which are then converted into tuples and emitted.
Since each thread is independently serving a shard, we cannot guarantee the ordering of the records in the replay state if we store the offsets. 
 
 
 
-----Original Message-----

**Comment by chaithu14** [comment]
This is the reason we are spooling all the data to HDFS and not just maintaining the offsets.

File: contrib/src/main/java/com/datatorrent/contrib/kinesis/AbstractKinesisOutputOperator.java (hunk 1 above)
Comment (chaithu14) [inline contrib/src/main/java/com/datatorrent/contrib/kinesis/AbstractKinesisOutputOperator.java]: Updated the code as per the comments and created new branch "KinesisEncoder" contains only the encoder support.

**Comment by chandnisingh** [comment]
There is a merge conflict with this pull request
Merge conflict in contrib/src/main/java/com/datatorrent/contrib/kinesis/AbstractKinesisInputOperator.java

**Comment by chaithu14** [comment]
Resolved the merge conflicts and opening a new pull request.

**Pull request closed without merging**